File: renderer/src/main/java/io/sparkled/renderer/util/ChannelPropPairUtils.java
package io.sparkled.renderer.util;

import io.sparkled.model.animation.ChannelPropPair;
import io.sparkled.model.animation.SequenceChannelEffects;
import io.sparkled.model.entity.SequenceChannel;
import io.sparkled.model.entity.StageProp;
import io.sparkled.model.util.GsonProvider;

import java.util.List;
import java.util.UUID;

import static java.util.stream.Collectors.toList;

/**
 * Pairs up sequence channels with their associated stage prop.
 */
public class ChannelPropPairUtils {

    private ChannelPropPairUtils() {
    }

    public static List<ChannelPropPair> makePairs(List<SequenceChannel> sequenceChannels, List<StageProp> stageProps) {
        return sequenceChannels.stream()
                .map(sc -> getPair(sc, stageProps))
                .collect(toList());
    }

    private static ChannelPropPair getPair(SequenceChannel sequenceChannel, List<StageProp> stageProps) {
        return new ChannelPropPair(convertChannelData(sequenceChannel), findStagePropByUuid(stageProps, sequenceChannel.getStagePropUuid()));
    }

    private static SequenceChannelEffects convertChannelData(SequenceChannel sequenceChannel) {
        return GsonProvider.get().fromJson(sequenceChannel.getChannelJson(), SequenceChannelEffects.class);
    }

    private static StageProp findStagePropByUuid(List<StageProp> stageProps, UUID uuid) {
        return stageProps.stream()
                .filter(sp -> sp.getUuid().equals(uuid))
                .findFirst()
                .orElse(null);
    }
}

File: renderer/src/main/kotlin/io/sparkled/renderer/util/ChannelPropPairUtils.kt
package io.sparkled.renderer.util

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import io.sparkled.model.SequenceChannelModel
import io.sparkled.model.StagePropModel
import io.sparkled.model.animation.ChannelPropPair
import io.sparkled.model.animation.SequenceChannelEffects

/**
 * Pairs up sequence channels with their associated stage prop.
 */
object ChannelPropPairUtils {

    fun makePairs(
        objectMapper: ObjectMapper,
        sequenceChannels: List<SequenceChannelModel>,
        stageProps: List<StagePropModel>
    ) = sequenceChannels.map { getPair(objectMapper, it, stageProps) }

    private fun getPair(
        objectMapper: ObjectMapper,
        sequenceChannel: SequenceChannelModel,
        stageProps: List<StagePropModel>,
    ): ChannelPropPair {
        return ChannelPropPair(
            convertChannelData(objectMapper, sequenceChannel),
            stageProps.first { sp -> sp.id == sequenceChannel.stagePropId }
        )
    }

    private fun convertChannelData(
        objectMapper: ObjectMapper,
        sequenceChannel: SequenceChannelModel,
    ): SequenceChannelEffects {
        return SequenceChannelEffects(
            effects = objectMapper.readValue(sequenceChannel.channelJson)
        )
    }
}
